homeassistant/components/media_player/squeezebox.py

```diff
@@ -60,129 +66,111 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
                       host, port, error)
         return False
 
-    # Combine it with port to allow multiple servers at the same host
-    key = "{}:{}".format(ipaddr, port)
-
     # Only add a media server once
-    if key in KNOWN_DEVICES:
+    if ipaddr in KNOWN_DEVICES:
         return False
-    KNOWN_DEVICES.append(key)
-
-    _LOGGER.debug("Creating LMS object for %s", key)
-    lms = LogitechMediaServer(host, port, username, password)
+    KNOWN_DEVICES.append(ipaddr)
 
-    if not lms.init_success:
+    _LOGGER.debug("Creating LMS object for %s", ipaddr)
+    lms = LogitechMediaServer(hass, host, port, username, password)
+    if lms is False:
         return False
 
-    add_devices(lms.create_players())
+    players = yield from lms.create_players()
+    yield from async_add_devices(players)
 
     return True
 
 
 class LogitechMediaServer(object):
     """Representation of a Logitech media server."""
 
-    def __init__(self, host, port, username, password):
+    def __init__(self, hass, host, port, username, password):
         """Initialize the Logitech device."""
+        self.hass = hass
         self.host = host
         self.port = port
         self._username = username
         self._password = password
-        self.http_port = self._get_http_port()
-        self.init_success = True if self.http_port else False
-
-    def _get_http_port(self):
-        """Get http port from media server, it is used to get cover art."""
-        http_port = self.query('pref', 'httpport', '?')
-        if not http_port:
-            _LOGGER.error("Failed to connect to server %s:%s",
-                          self.host, self.port)
-        return http_port
 
+    @asyncio.coroutine
     def create_players(self):
-        """Create a list of SqueezeBoxDevices connected to the LMS."""
-        players = []
-        count = self.query('player', 'count', '?')
-        for index in range(0, int(count)):
-            player_id = self.query('player', 'id', str(index), '?')
-            player = SqueezeBoxDevice(self, player_id)
-            players.append(player)
-        return players
-
-    def query(self, *parameters):
+        """Create a list of devices connected to LMS."""
+        result = []
+        data = yield from self.async_query('players', 'status')
+
+        for players in data['players_loop']:
+            player = SqueezeBoxDevice(
+                self, players['playerid'], players['name'])
+            yield from player.async_update()
+            result.append(player)
+        return result
+
+    @asyncio.coroutine
+    def async_query(self, *parameters, player=""):
         """Send request and await response from server."""
-        response = self.get(' '.join(parameters))
-        response = response.split(' ')[-1].strip()
-        response = urllib.parse.unquote(response)
-
+        response = yield from self._query(parameters, player)
```

Comment:
Why is there a secondary coroutine here? Can this logic be consolidated into `async_query`?